### ISSUE TITLE
Missing $count/inlinecount support for OData provider

### ISSUE THREAD

**Issue by whtan98** [ghiscoding/Angular-Slickgrid#287]
## I'm submitting a Bug report
The currentOdataQueryBuilderService does not support the following option:
1) $count (OData4)
 2) $inlinecount (OData2)
### Your Environment
| Software          | Version(s) |
| ----------------- | ---------- |
| Angular           | 7/8        | 
| Angular-Slickgrid | 2.10.5         | 

### Context
OdataQueryBuilderService does not support "enableCount" option

### Possible Solution

Suggest to add the following to OdataQueryBuilderService .buildQuery()
```
 /*
    * Build the OData query string from all the options provided
    * @return string OData query
    */
  buildQuery(): string {
    if (!this._odataOptions) {
      throw new Error('Odata Service requires certain options like "top" for it to work');
    }
    this._odataOptions.filterQueue = [];
    const queryTmpArray = [];

    // When enableCount is true
    if (this._odataOptions.enableCount === true) {
      if (this._odataOptions.version === 4) {
        queryTmpArray.push(`$count=true`);
      } else if (this._odataOptions.version <= 3 ) {
        queryTmpArray.push(`$inlinecount=allpages`);
      }
    }
   //more...
```

**Comment by ghiscoding** [comment]
Oh I wasn't aware that OData provides total count in their query, I thought we always had to provide that ourselves in the .Net Controller. I wonder if the boolean flag is really necessary but at the same time, we could disable the pagination so I guess the flag could be useful there. 

I'll take a look at that later, I thought at first this might be a breaking change but after a quick look, I guess not since user can pass whatever variable name. 

Can you provide a quick sample of dataset returned after this query change? I don't have access to OData server anymore

**Comment by whtan98** [comment]
Yes. See the following V3/V4 odata services 

1) OData3
curl 'https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$top=2&$inlinecount=allpages' -H "Accept: application/json;"
 
2)OData4 
curl 'https://services.odata.org/V4/Northwind/Northwind.svc/Customers?$top=2&$count=true'  -H "Accept: application/json;"

**Comment by ghiscoding** [comment]
Thanks, however if I run those queries you sent, it shows the count in a strange variable. Is `"@odata.count"` a generic name and you can change it or is that the commonly used name and is what you get on your side as well? I need to have an answer on this since this will influence the Pagination.

```json
{
	"@odata.context":"https://services.odata.org/V4/Northwind/Northwind.svc/$metadata#Customers",
	"@odata.count":91,
	"value":[{"CustomerID":"ALFKI","CompanyName":"Alfreds Futterkiste","ContactName":"Maria Anders","ContactTitle":"Sales Representative","Address":"Obere Str. 57","City":"Berlin","Region":null,"PostalCode":"12209","Country":"Germany","Phone":"030-0074321","Fax":"030-0076545"},{"CustomerID":"ANATR","CompanyName":"Ana Trujillo Emparedados y helados","ContactName":"Ana Trujillo","ContactTitle":"Owner","Address":"Avda. de la Constituci\u00f3n 2222","City":"M\u00e9xico D.F.","Region":null,"PostalCode":"05021","Country":"Mexico","Phone":"(5) 555-4729","Fax":"(5) 555-3745"}]
}
```

**Comment by whtan98** [comment]
I got the same name as well.
It is the common name as defined by the spec. See
https://docs.oasis-open.org/odata/odata-json-format/v4.01/csprd05/odata-json-format-v4.01-csprd05.html#_Toc14363451

**Comment by ghiscoding** [comment]
@whtan98 
With this change, which I'm currently working on adding, what is your response structure?

is it
```ts
{
   "@odata.count":91,
   "value":[{"CustomerID":"ALFKI","CompanyName":"Alfreds Futterkiste","ContactName":"Maria Anders","ContactTitle":"Sales Representative","Address":"Obere Str. 57","City":"Berlin","Region":null,"PostalCode":"12209","Country":"Germany","Phone":"030-0074321","Fax":"030-0076545"},{"CustomerID":"ANATR","CompanyName":"Ana Trujillo Emparedados y helados","ContactName":"Ana Trujillo","ContactTitle":"Owner","Address":"Avda. de la Constituci\u00f3n 2222","City":"M\u00e9xico D.F.","Region":null,"PostalCode":"05021","Country":"Mexico","Phone":"(5) 555-4729","Fax":"(5) 555-3745"}]
}
```

or

**Comment by ghiscoding** [comment]
I created the PR #291, if you could review it and answer the question in [previous post](https://github.com/ghiscoding/Angular-Slickgrid/issues/287#issuecomment-534748027) before I merge the code. The PR has a few files because I also updated Jest unit tests and Cypress E2E tests.

Note that I changed your code a bit because `version` is not a required property and when it's not provided it defaults to OData v2, here's the new code. 
```ts
if (this._odataOptions.enableCount === true) {
      const countQuery = (this._odataOptions.version >= 4) ? '$count=true' : '$inlinecount=allpages';
      queryTmpArray.push(countQuery);
}
```

**Comment by whtan98** [comment]
> @whtan98
> With this change, which I'm currently working on adding, what is your response structure?
> 
> is it (I also saw that it's `@odata.count` in v4 but `odata.count` in v2)
> 

I got exactly the same observations as yours.

However,  things may get tricky though, after reading the following  excerpts from ODATAv4 spec https://docs.oasis-open.org/odata/odata-json-format/v4.01/csprd05/odata-json-format-v4.01-csprd05.html#_Toc14363451

See picture below:
![odata](https://user-images.githubusercontent.com/13287569/65575849-529a0680-dfa3-11e9-924f-9e05028b71c9.png)

**Comment by ghiscoding** [comment]
From what you pointed out, that seems to mean that the result will be `@count` instead of `@data.count` because we always use a GET without the version in the header. I think that in any case, you have to deal with the total count yourself in the Component, so that would work out anyway since you'll use the name that works for you. As long as the code change for `$inlinecount` and `$count` is still what you provided then I think we are fine. Can you validate that on your side

**Comment by whtan98** [comment]
Agreed. The count control  info should be handled separately.

BTW, below are my findings for the default name:
1) V3 is "odata.count"
2) V4 is "@odata.count"

As the odata service always specify the ODATA-Version in the RESPONSE HTTP header

See attached curl output
[output.txt](https://github.com/ghiscoding/Angular-Slickgrid/files/3653256/output.txt)

Will let you know once I tested the changes.

Many thanks

**Comment by ghiscoding** [comment]
Oh I thought the text you sent for only for the Request Header but now I see it's for both Request & Response Headers. Anyhow, my lib won't explicitly change anything in the Request Header, the only thing my lib does is to build the OData query string, the user (you) do whatever you want to with the Request afterward (meaning you can choose to add the OData version in your Request if you wish but I assume most people won't because again my lib doesn't handle the http call itself).

Also, I looked back at the OData Wiki and I realized that the callback, which has to change the pagination total count property, was not shown in the [OData code Wiki](https://github.com/ghiscoding/Angular-Slickgrid/wiki/OData#code) so I updated the Wiki, basically the following
```ts
getCustomerCallback(response) {
    // totalItems property needs to be filled for pagination to work correctly
    // however we need to force Angular to do a dirty check, doing a clone object will do just that
    this.gridOptions.pagination.totalItems = response.totalRecordCount;
    this.gridOptions = Object.assign({}, this.gridOptions);

    this.dataset = response.items as Customer[];
  }
```
However the OData example always had that piece of code.. and again it's up to the user (you) to handle the total count returned from the callback and update the pagination count (else the pagination will be broken).

**Comment by whtan98** [comment]
> Can you validate that it works on your side. I'd like to release a new version this week. Thanks
It works correctly for me. Many thanks for the great effort

**Comment by ghiscoding** [comment]
@whtan98 
Can you share the code of your `postProcess` callback? I might do improvement in the near future to auto-refresh internally (that is something I already do with the GraphQL service) but I need to make sure it's always following a certain structure.

**Comment by whtan98** [comment]
I'm afraid I don't have any proper code for postProcess  yet, as I still very much on exploring how to make use of Angular SlickGrid

**Comment by ghiscoding** [comment]
Guess I'll wait to do that next enhancement then, I need someone with access to OData. Thanks

**Comment by ghiscoding** [comment]
Now released under version [2.11.0](https://github.com/ghiscoding/Angular-Slickgrid/releases/tag/v2.11.0)

**Comment by whtan98** [comment]
Great work. Thank you !
